File: Dynamic_Programming/C/coin_exchange2.c
/*
Author:         Kumar Shanu
Problem:        Find out the total number of ways you can make the change of
                        given amount using given coins.
Problem link:   https://leetcode.com/problems/coin-change-2/

Dynamic Programming
*/

#include <stdio.h>
#include <string.h>

int findchange(int coins[], int n, int amount)
{
    if (amount == 0)
        return 1;
    if (n == 0)
        return 0;

    int arr[n][amount + 1];

    memset(arr, 0, sizeof(arr));

    for (int j = 0; j <= amount; j++)
        arr[0][j] = (int)(j % coins[0] == 0);

    for (int i = 1; i < n; i++)
    {
        arr[i][0] = 1;
        for (int j = 0; j <= amount; j++)
        {
            if (coins[i] > j)
                arr[i][j] = arr[i - 1][j];
            else
                arr[i][j] = arr[i - 1][j] + arr[i][j - coins[i]];
        }
    }

    return arr[n - 1][amount];
}

int main()
{
    int coins[] = {2, 3, 5, 10}, w = 15;

    printf("%d\n", findchange(coins, 4, w));

    return 0;
}

File: Dynamic_Programming/python/coin_exchange2.py
"""
Author:         Kumar Shanu
Problem:        Find out the total number of ways you can make the change of
                        given amount using given coins.
Problem link:   https://leetcode.com/problems/coin-change-2/

Dynamic Programming
"""


class Solution:
    def change(self, amount: int, coins: "list[int]") -> int:
        n, m = len(coins), amount+1
        if amount == 0:
            return 1
        if n == 0:
            return 0
        coins.sort()
        arr = [[0 for i in range(m)] for j in range(n)]

        for j in range(m):
            arr[0][j] = int(j % coins[0] == 0)

        for i in range(1, n):
            arr[i][0] = 1
            for j in range(m):
                if coins[i] > j:
                    arr[i][j] = arr[i-1][j]
                else:
                    arr[i][j] = arr[i-1][j]+arr[i][j-coins[i]]

        print(arr)
        return arr[-1][-1]


if __name__ == "__main__":
    coins = [2, 3, 5, 10]
    w = 15
    s = Solution().change(w, coins)
    print(s)
